lib/simplecov/combine/results_combiner.rb

```diff
@@ -33,17 +33,57 @@ def combine(*results)
       #
       # @return [Hash]
       #
-      def combine_result_sets(result_a, result_b)
-        results_files = result_a.keys | result_b.keys
+      def combine_result_sets(combined_results, result)
+        unless correct_format?(result)
+          warn_wrong_format
+          return combined_results
+        end
+
+        results_files = combined_results.keys | result.keys
 
-        results_files.each_with_object({}) do |file_name, combined_results|
-          combined_results[file_name] = combine_file_coverage(
-            result_a[file_name],
-            result_b[file_name]
+        results_files.each_with_object({}) do |file_name, file_combination|
+          file_combination[file_name] = combine_file_coverage(
+            combined_results[file_name],
+            result[file_name]
           )
         end
       end
 
+      # We might start a run of a new simplecov version with a new format stored while
+      # there is still a recent file like this lying around. If it's recent enough (
+      # see merge_timeout) it will end up here. In order not to crash against this
+      # we need to do some basic checking of the format of data we expect and
+      # otherwise ignore it. See #820
+      #
+      # Currently correct format is:
+      # { symbol_file_path => {coverage_criterion => coverage_date}}
+      #
+      # Internal use/reliance only.
+      def correct_format?(result)
+        result.empty? || matches_current_format?(result)
+      end
+
+      def matches_current_format?(result)
+        # I so wish I could already use pattern matching
+        key, data = result.first
```

Comment:
this whole checking (and the other checks) feel kinda clunky. I didn't know of a better way to handle it right now though :(